### PR TITLE
Reorder expression operators

### PR DESCRIPTION
**Pull request by roya0045**
## Description

This is a proposal to reduce the importance of the concatenate operator in expressions.

Currently the '||' operator is evaluated first. This means that any numerical value is converted to text and bricks any related mathematical operator.

This fix proposes to evaluate mathematical operators first and then do the one-way string transformation.

### CODE REVIEW

**Comment by roya0045** [comment]
@m-kuhn @nyalldawson A review would be appreciated to validate this changes. I cannot expect any downsides from this change, but I may be wrong.

**Comment by nyalldawson** [comment]
Seems there's ambiguity here, and eg sqlite considers || as higher precedence (https://www.sqlite.org/lang_expr.html#operators) while postgres treats it as lower precedence (https://www.postgresql.org/docs/8.0/sql-syntax.html#SQL-PRECEDENCE)

@wonder-sk do you have thoughts here?

**Comment by nyalldawson** [comment]
My gut feeling is that while I'd personally expect || to be lower precedence, given that there's no standard here we should play it safe and not change the order (which could potentially break existing projects) (especially given there's an easy user fix of manually inserting parenthesis where desired)

**Comment by roya0045** [comment]
> My gut feeling is that while I'd personally expect || to be lower precedence, given that there's no standard here we should play it safe and not change the order (which could potentially break existing projects) (especially given there's an easy user fix of manually inserting parenthesis where desired)

I'd suggest the opposite. If everyone agrees that this operator should not have preference over mathematical operators, than having it the current way maintain an unintuitive behaviour. Though I would like feedback from as many people as possible.

**Comment by nyalldawson** [comment]
>. If everyone agrees that this operator should not have preference over mathematical operators

There clearly isn't a consensus here though, hence the difference between sqlite vs postgres behaviour...

**Comment by roya0045** [comment]
Maybe for sql provider the case is distinct. 

Though you state that giving priority to other operator rather than concat may break some things. Given the fact that to mathematical operation can occur after a concatenation, I don't think there are valid expressions where mathematical operator and concatenate exist at the same level, as concat exal first and makes any further mathematical operating invalidating the expression.

Expressions using the workaround are already safe. I honestly don't think this change will have any negative effect.

I can open a thread in the dev mailing list to raise this point for discussion.

**Comment by roya0045** [comment]
@uclaros @nirvn @gioman Curious to know your opinion on the expected behaviour of concat on a level with mathematical operators,

**Comment by uclaros** [comment]
At first I though *that's safe and intuitive, lets change it*. On second thought, it seems more complicated.

> Given the fact that to mathematical operation can occur after a concatenation, I don't think there are valid expressions where mathematical operator and concatenate exist at the same level, as concat exal first and makes any further mathematical operating invalidating the expression.

Since `+` acts as a `||` between strings and `-` implicitly tries to convert the strings to numerics (ie `'4' + '2' = '42'` but `'4' + '2' - '1' = 41` ) then it could be possible that someone uses expressions that will indeed break by the change.

Having said that, I agree that it seems more intuitive and useful to follow the postgres way. Maybe mark it as a todo for qgis 4?

**Comment by NathanW2** [comment]
Changing it would also break the results of existing projects and
expressions correct?  That seems like a major risk to me.

**Comment by nyalldawson** [comment]
@NathanW2 

> Changing it would also break the results of existing projects and
expressions correct?  That seems like a major risk to me.

That's my concern too -- I'd be ok with that if this was clearly a bug, but given that we currently the sqlite/mysql approach it's debatable if this is an issue or not in the first place. That's why I'm -0 to merge. 

But let's wait for @wonder-sk's feedback, given that he originally set the precedence order.

**Comment by roya0045** [comment]
I honestly could not come up with a valid use case that would benefit from the current behaviour and not the proposed one. As soon as I have a case I'll be more willing to back off or post-pone this.

**Comment by nyalldawson** [comment]
(this PR gains the new "tricky!!" label, which means "requires very careful consideration and review prior to merge"!)

**Comment by wonder-sk** [comment]
When initially working on the QgsExpression, I was mostly drawing inspiration from SQLite and PostgreSQL - comparing how they operated and figuring out what would make most sense in our simple SQL-like language. It looks like I ended up copying the precedence of operators from SQLite - whenever I had doubts I was trying to make things compatible at least with one of the engines :-)

Now, with string concatenation, that's a total hell in SQL - every single engine implements things differently, with different rules of handling numbers, nulls and so on, so the operator precedence is just icing on the cake :-) To add to Nyall's list of precedences in other implementations:
- Oracle seems to have the same precedence for + and ||  (https://docs.oracle.com/cd/B19306_01/server.102/b14200/operators001.htm)
- MySQL seems to have nearly the lowest possible precedence for || (https://dev.mysql.com/doc/refman/8.0/en/operator-precedence.html)
- MS SQL does not seem to have || operator, there's + operator for string concatenation

To summarize it, I think string concatenation operator does not have a widely accepted precedence, therefore increasing or decreasing it probably won't make things better or worse... just different :-) And then we only have the risk of breaking expressions of some users, which I would hate to do. So I would be in favor of keeping things as they are now...

**Comment by roya0045** [comment]
Closing for now then. Maybe worth considering for 4.0.
